common.go

```diff
@@ -0,0 +1,96 @@
+package cose
+
+import (
+	"errors"
+	"fmt"
+)
+
+// intOrStr is a value that can be either an int or a tstr when serialized to
+// CBOR.
+type intOrStr struct {
+	intVal   int64
+	strVal   string
+	isString bool
+}
+
+func newIntOrStr(v interface{}) *intOrStr {
+	var ios intOrStr
+	if err := ios.Set(v); err != nil {
+		return nil
+	}
+	return &ios
+}
+
+func (ios intOrStr) Int() int64 {
+	return ios.intVal
+}
+
+func (ios intOrStr) String() string {
+	if ios.IsString() {
+		return ios.strVal
+	}
+	return fmt.Sprint(ios.intVal)
+}
+
+func (ios intOrStr) IsInt() bool {
+	return !ios.isString
+}
+
+func (ios intOrStr) IsString() bool {
+	return ios.isString
+}
+
+func (ios intOrStr) Value() interface{} {
+	if ios.IsInt() {
+		return ios.intVal
+	}
+
+	return ios.strVal
+}
+
+func (ios *intOrStr) Set(v interface{}) error {
+	switch t := v.(type) {
+	case int64:
```

Comment:
Can `v` be a `int8`, `int16`, `int32`, or their unsigned counterparts?

Reply:
It cannot be be any of those. This is an unexported type used solelly for CBOR (un)marshalling.`fxamacker/cbor/v2` only unmarshals ints as either `uint64` or `int64`, and because we're setting `IntDecCovertSigned` here: https://github.com/veraison/go-cose/blob/main/cbor.go#L43, it will only ever be `int64`.